File: common/netstack_rs/src/cxx/wrapper.cpp
/*
 * Copyright (C) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "wrapper.h"

#include <cstring>
#include <memory>

#include "http_client_error.h"
#include "net_conn_client.h"
#include "wrapper.rs.h"

namespace OHOS::Request {
using namespace OHOS::NetStack::HttpClient;
using namespace OHOS::NetManagerStandard;

void OnCallback(const std::shared_ptr<HttpClientTask> &task, rust::Box<CallbackWrapper> callback)
{
    CallbackWrapper *raw_ptr = callback.into_raw();
    auto shared = std::shared_ptr<CallbackWrapper>(
        raw_ptr, [](CallbackWrapper *ptr) { rust::Box<CallbackWrapper>::from_raw(ptr); });
    task->OnSuccess([shared](const HttpClientRequest &request, const HttpClientResponse &response) {
        shared->on_success(request, response);
    });
    task->OnFail([shared](const HttpClientRequest &request, const HttpClientResponse &response,
                     const HttpClientError &error) { shared->on_fail(request, response, error); });
    task->OnCancel([shared](const HttpClientRequest &request, const HttpClientResponse &response) {
        shared->on_cancel(request, response);
    });
    auto weak = task->weak_from_this();
    task->OnDataReceive([shared, weak](const HttpClientRequest &, const uint8_t *data, size_t size) {
        auto httpTask = weak.lock();
        if (httpTask != nullptr) {
            shared->on_data_receive(httpTask, data, size);
        }
    });
    task->OnProgress([shared](const HttpClientRequest &, u_long dlTotal, u_long dlNow, u_long ulTotal, u_long ulNow) {
        shared->on_progress(dlTotal, dlNow, ulTotal, ulNow);
    });
};

void GetPerformanceInfo(const HttpClientResponse &response, RustPerformanceInfo &performance)
{
    OHOS::NetStack::HttpClient::PerformanceInfo cpp_perf = response.GetPerformanceTiming();
    performance.set_dns_timing(cpp_perf.dnsTiming);
    performance.set_connect_timing(cpp_perf.connectTiming);
    performance.set_tls_timing(cpp_perf.tlsTiming);
    performance.set_first_send_timing(cpp_perf.firstSendTiming);
    performance.set_first_receive_timing(cpp_perf.firstReceiveTiming);
    performance.set_total_timing(cpp_perf.totalTiming);
    performance.set_redirect_timing(cpp_perf.redirectTiming);
}
} // namespace OHOS::Request

File: common/utils/src/dataability/wrapper.rs
// Copyright (C) 2024 Huawei Device Co., Ltd.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Foreign Function Interface (FFI) bindings for the request system.
//!
//! This module provides FFI bindings to C++ code through the CXX bridge, enabling
//! interaction with native APIs from Rust. It includes types, enums, and functions
//! for accessing system services, logging, and storage functionality.

#![allow(unused)]

use cxx::SharedPtr;
pub use ffi::*;

// CXX bridge module for FFI bindings to C++ code
#[cxx::bridge(namespace = "OHOS::Request")]
mod ffi {

    // C++ functions and types exposed to Rust
    unsafe extern "C++" {
        include!("request_utils_dataability.h");
        
        #[namespace = "OHOS::AbilityRuntime"]
        type Context = request_utils::wrapper::Context;

        fn DataAbilityOpenFile(context: &SharedPtr<Context>, path: &CxxString) -> i32;
    }
}